Magnum Player

mosra opened the pull request:
This will become an extended version of the Viewer example with the end goal being an ability to view and play any 3D file you throw at it. 

The initial goals, in scope of this PR, are:

- [x] Basic loading of scene files in scope that the Viewer example did
- [x] Camera controls better than the Viewer example (panning, depth-aware rotation -- might reuse mosra/magnum-examples#46)
- [x] Mention the app in Magnum's `doc/utilities.dox`
- [x] Mention the app in in the viewer example docs
- [x] Screenshot in the docs
- [x] Loading of animations and playing them back in a loop
  - [x] Ability to pause them
  - [x] Frame stepping
- [x] Web port
  - [x] Upgrading the default web template
  - [x] DPI awareness
  - [x] Drag&drop instead of having a single hardcoded file
  - [x] ~~Ability to specify aspect ratio of the view~~ Ability to full-size the view
  - [x] Ability to D&D multiple files
- [x] ~~Screenshot saving~~ postponed
  - [x] Enable JPEG export once done (mosra/magnum-plugins#26)
- [x] Resolve the issue with multiple lights changing alpha desired mask (the gltf alpha example fails)
- [x] ~~Some default scene for the web view (animated cube at least)~~ postponed
- [x] ~~Figure out the depth issues~~ postponed
- [x] Enable MSAA
- [x] ~~Some better scene info (a modal) .. also cam count etc.~~ postponed

mosra committed:
Bootstrap the magnum-player application.

Initial code is a plain copy of the Viewer example.

mosra committed:
player: switch to the new TRS transformation implementation.

And if possible, import the initial TRS properties separately. Makes it
better for handling animations.

mosra committed:
player: initial support for loading and playing back animations.

mosra committed:
player: enable alpha masking for cheap&ugly alpha cutoff.

mosra committed:
player: import the model-defined camera, if possible.

mosra committed:
player: be nice to GLES2 and WebGL.

mosra committed:
player: extract just the matrix rotation/scaling part altogether.

Better than crashing during an animation.

mosra committed:
player: prepare the data to be reloadable.

For drag & drop.

mosra committed:
player: initial port to Emscripten.

The file will be specified through drag&drop, so this is not doing
anything at the moment.

Squareys commented on src/player/player.html:
```diff
@@ -0,0 +1,19 @@
+  </head>
+  <body>
+    <h1>Magnum glTF Player</h1>
+    <p style="max-width: 640px; text-align: center; margin-left: auto; margin-right: auto;">Drag &amp; drop your <tt>*.glb</tt> (or an <em>embedded</em> <tt>*.gltf</tt>) file into the grey rectange below to play it. Refresh the browser if it gets stuck. I'm not responsible for anything :P</p>
```
"rectange" should be "rectangle" 😉

mosra committed:
package: enable the player app in all packages and CIs.

It needs Ui, which doesn't work on WebGL 1 / ES2.

mosra committed:
player: implement drag&drop for Emscripten.

mosra committed:
player: support spline-interpolated animations.

mosra committed:
player: basic three-point light setup.

mosra committed:
player: bigger FoV.

mosra committed:
player: don't kill the alpha mask with specular color.

mosra committed:
player: don't just ignore meshes without normals.

mosra committed:
player: deduplicate animation import code a bit.

mosra committed:
player: initial UI.

mosra committed:
player: don't capture the keyboard by default.

mosra committed:
player: display also the filename, remove now-redundant debug info.

mosra committed:
player: "relayout" the UI on window size change.

Well, basically it's recreating it from scratch. Sorry, ugly, verbose
and slow. Needs to be done properly later.

mosra committed:
player: a button to switch to full-size view and back.

mosra committed:
player: rudimentary support for blend and alpha mask.

mosra committed:
player: ability to control glTF animation clip merging.

mosra committed:
player: adapt to Magnum changes.

mosra committed:
player: reorganize the privates a bit.

mosra committed:
player: depth-aware mouse interaction.

Reused from the Mouse Interaction example.

mosra committed:
player: make the lights relative to the camera again.

mosra committed:
player: port the depth-aware mouse interaction to WebGL.

Again just a copy from the Mouse Interaction example.

mosra committed:
player: try to use MSAA by default.

But only 2x on HiDPI and allow the users to override that.

mosra committed:
player: support dropping multiple files.

Apparently hierarchy is not preserved with browser D&D so if a file in
subdir is not found, we try again using just the filename.

mosra committed:
player: initialize the animation time destination.

Otherwise the animation time might not get displayed when the
destination random value would be matching animation progress.

mosra committed:
player: seeking UI and controls.

mosra committed:
Ui: include proper configure header.

Looks like static build on Windows was never working properly.

mosra committed:
player: wrap up the docs, add a screenshot.

mosra committed:
Ui: add an image for the magnum-ui-gallery app.
